From 9b360f9e1b27fe7091d867de39bf5e2ca88f4e09 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Fri, 11 Jun 2021 13:59:44 +0200
Subject: [PATCH] PONRTSYS-9035: Remove dma_map_single()

Instead of allocating the memory with dma_alloc_noncoherent() and
ignoring the physical address and using dma_map_single() to get the
physical address, just use the physical address returned by
dma_alloc_noncoherent(). The function buf_addr_adjust() is extended to
also align the physical address.
dma_alloc_noncoherent() returns normal DMA mapped able memory.
---
 drivers/net/ethernet/lantiq/cqm/cqm_common.c        |  6 ++++++
 drivers/net/ethernet/lantiq/cqm/cqm_common.h        |  2 ++
 drivers/net/ethernet/lantiq/cqm/grx500/cbm.c        |  4 ++++
 drivers/net/ethernet/lantiq/cqm/prx300/cqm.c        | 15 +++------------
 drivers/net/ethernet/lantiq/cqm/prx300/cqm_config.h |  1 +
 5 files changed, 16 insertions(+), 12 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/cqm/cqm_common.c b/drivers/net/ethernet/lantiq/cqm/cqm_common.c
index 795733898ea0..71aedf55c89f 100644
--- a/drivers/net/ethernet/lantiq/cqm/cqm_common.c
+++ b/drivers/net/ethernet/lantiq/cqm/cqm_common.c
@@ -18,19 +18,25 @@ static void __iomem *cqm_ls_addr_base;
 
 void buf_addr_adjust(unsigned int buf_base_addr,
 		     unsigned int buf_size,
+		     dma_addr_t buf_dma_addr,
 		     unsigned int *adjusted_buf_base,
 		     unsigned int *adjusted_buf_size,
+		     dma_addr_t *adjusted_dma_addr,
 		     unsigned int align)
 {
 	unsigned int base;
 	unsigned int size;
+	dma_addr_t dma_addr;
 
 	pr_info("0x%x 0x%x 0x%x\n", buf_base_addr, buf_size, align);
 	base = ALIGN(buf_base_addr, align);
 	size = buf_base_addr + buf_size - base;
+	dma_addr = ALIGN(buf_dma_addr, align);
 
 	*adjusted_buf_base = base;
 	*adjusted_buf_size = size;
+	if (adjusted_dma_addr)
+		*adjusted_dma_addr = dma_addr;
 	pr_info("0x%x 0x%x\n", base, size);
 }
 
diff --git a/drivers/net/ethernet/lantiq/cqm/cqm_common.h b/drivers/net/ethernet/lantiq/cqm/cqm_common.h
index d0357f4e127c..169f9ee9129b 100644
--- a/drivers/net/ethernet/lantiq/cqm/cqm_common.h
+++ b/drivers/net/ethernet/lantiq/cqm/cqm_common.h
@@ -135,8 +135,10 @@ static inline int get_is_bit_set(u32 flags)
 }
 
 void buf_addr_adjust(unsigned int buf_base_addr, unsigned int buf_size,
+		     dma_addr_t buf_dma_addr,
 		     unsigned int *adjusted_buf_base,
 		     unsigned int *adjusted_buf_size,
+		     dma_addr_t *adjusted_dma_addr,
 		     unsigned int align);
 int cbm_linearise_buf(struct sk_buff *skb, struct cbm_tx_data *data,
 		      int buf_size, u32 new_buf);
diff --git a/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c b/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c
index 0f8824148673..d8e0ca5be9a0 100644
--- a/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c
+++ b/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c
@@ -881,13 +881,17 @@ static void init_fsqm(void)
 {
 	buf_addr_adjust((u32)std_buf_base,
 			CONFIG_CBM_STD_BUF_POOL_SIZE,
+			0,
 			&g_cbm_buff.std_buf_addr,
 			&g_cbm_buff.std_buf_size,
+			NULL,
 			g_cbm_buff.std_frm_size);
 	buf_addr_adjust((u32)jbo_buf_base,
 			CONFIG_CBM_JBO_BUF_SIZE,
+			0,
 			&g_cbm_buff.jbo_buf_addr,
 			&g_cbm_buff.jbo_buf_size,
+			NULL,
 			g_cbm_buff.jbo_frm_size);
 
 	g_cbm_buff.std_frm_num =
diff --git a/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c b/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c
index 5f47012f45f8..b993f05a9e79 100644
--- a/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c
+++ b/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c
@@ -3918,8 +3918,10 @@ static int pool_config(struct platform_device *pdev, int p, u32 buf_size,
 	p_ptr->buf_frm_num = buf_num;
 	buf_addr_adjust((u32)cqm_ctrl->bm_buf_base[p],
 			size,
+			cqm_ctrl->dma_hndl_p[p],
 			&p_ptr->pool_start_low,
 			&adj_size,
+			&p_ptr->pool_dma,
 			p_ptr->buf_frm_size);
 	p_ptr->pool_end_low = p_ptr->pool_start_low + adj_size;
 	p_ptr->buf_frm_size_reg = get_bufsize(buf_size);
@@ -3931,8 +3933,6 @@ static int bm_init(struct platform_device *pdev)
 	struct pp_bmgr_pool_params p_params;
 	u8 i, j;
 	int result;
-	u32 buf_size;
-	u32 start_low;
 	u8  req_pool_id = PP_BM_INVALID_POOL_ID;
 	u16 req_policy_id = PP_BM_INVALID_POLICY_ID;
 	s32 ret;
@@ -3951,16 +3951,7 @@ static int bm_init(struct platform_device *pdev)
 	for (i = 0; i < cqm_ctrl->num_pools; i++) {
 		p_params.num_buffers = bm_pool_conf[i].buf_frm_num;
 		p_params.size_of_buffer = bm_pool_conf[i].buf_frm_size;
-		start_low = bm_pool_conf[i].pool_start_low;
-		buf_size = p_params.num_buffers * p_params.size_of_buffer;
-		buf_size *= sizeof(u32);
-		p_params.base_addr_low = dma_map_single(cqm_ctrl->dev,
-							(void *)start_low,
-							buf_size,
-							DMA_TO_DEVICE);
-		/* The buffers are reserved for BM module.
-		 * no need to call dma_unmap_single.
-		 */
+		p_params.base_addr_low = bm_pool_conf[i].pool_dma;
 		p_params.base_addr_high = 0;
 		p_params.flags = POOL_ENABLE_FOR_MIN_GRNT_POLICY_CALC;
 		req_pool_id = PP_BM_INVALID_POOL_ID;
diff --git a/drivers/net/ethernet/lantiq/cqm/prx300/cqm_config.h b/drivers/net/ethernet/lantiq/cqm/prx300/cqm_config.h
index 7bad13c7290c..4ecf40bc7f17 100644
--- a/drivers/net/ethernet/lantiq/cqm/prx300/cqm_config.h
+++ b/drivers/net/ethernet/lantiq/cqm/prx300/cqm_config.h
@@ -128,6 +128,7 @@ struct cqm_bm_pool_config {
 	unsigned int pool_start_high;
 	unsigned int pool_end_low;
 	unsigned int pool_end_high;
+	dma_addr_t pool_dma;
 	unsigned int segment_mask;
 	unsigned int offset_mask;
 	unsigned int pool;
